From 28abd793dd2fd52684fa41e8fc109df5a373220b Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Fri, 25 Dec 2020 13:32:09 +0700
Subject: [PATCH 23/23] rzg: g2e: add support for identifying board revision

This commit is created to support identifying EK874 board revision
by checking value of GP5_19.
In latest EK874 Revisions, this pin is pulled down with an external
4.7kOhm resistor. Therefore, when we enable internal pull-up then
setting general input mode for this pin, we will find the different
value of GP5_19 in each revisions:
- Rev.C and the older: 1
- The latest Rev.E: 0

And also enable support pins for eMMC (SDHI3-IF) for the latest EK874
revisions.

Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 drivers/renesas/rzg/board/board.c          |  8 +++++-
 drivers/renesas/rzg/pfc/G2E/pfc_init_g2e.c | 44 ++++++++++++++++++++++++++++++
 plat/renesas/rzg/bl2_plat_setup.c          | 10 +++++--
 3 files changed, 59 insertions(+), 3 deletions(-)

diff --git a/drivers/renesas/rzg/board/board.c b/drivers/renesas/rzg/board/board.c
index d8d7eb0..7e26437 100644
--- a/drivers/renesas/rzg/board/board.c
+++ b/drivers/renesas/rzg/board/board.c
@@ -43,6 +43,10 @@
 #define HH_ID	{ 0x40U, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU }
 #define EK_ID	{ 0x10U, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU }
 
+#if (RZG_LSI == RZG_G2E)
+extern char ek874_board_rev;
+#endif /* RZG_LSI == RZG_G2E */
+
 const char *g_board_tbl[] = {
 	[BOARD_HIHOPE_RZG2M]	= "HiHope RZ/G2M",
 	[BOARD_HIHOPE_RZG2N]	= "HiHope RZ/G2N",
@@ -65,7 +69,7 @@ int32_t rzg_get_board_type(uint32_t *type, uint32_t *rev)
 	uint32_t reg, boardInfo, read_rev;
 #elif (RZG_LSI == RZG_G2H)
 	uint32_t boardInfo;
-#else
+#elif (RZG_LSI != RZG_G2E)
 	uint32_t read_rev;
 #endif
 
@@ -110,6 +114,8 @@ get_type:
 #elif (RZG_LSI == RZG_G2H)
 	boardInfo = mmio_read_32(GPIO_INDT5) & (GP5_19_BIT |GP5_21_BIT);
 	*rev = (((boardInfo & GP5_19_BIT) >> 14) | ((boardInfo & GP5_21_BIT) >> 17)) + 0x30;
+#elif (RZG_LSI == RZG_G2E)
+	*rev = ek874_board_rev;
 #else
 	read_rev = (uint8_t)(board_id & BOARD_REV_MASK);
 	*rev = board_tbl[*type][read_rev];
diff --git a/drivers/renesas/rzg/pfc/G2E/pfc_init_g2e.c b/drivers/renesas/rzg/pfc/G2E/pfc_init_g2e.c
index 31d07f5..0257847 100644
--- a/drivers/renesas/rzg/pfc/G2E/pfc_init_g2e.c
+++ b/drivers/renesas/rzg/pfc/G2E/pfc_init_g2e.c
@@ -9,6 +9,7 @@
 #include "pfc_init_g2e.h"
 #include "rzg_def.h"
 #include "../pfc_regs.h"
+#include "micro_delay.h"
 
 /* PFC */
 #define GPSR0_SDA4		BIT(17)
@@ -287,6 +288,10 @@
 #define MOD_SEL1_SSI9_A		((uint32_t)0U << 4U)
 #define MOD_SEL1_SSI9_B		((uint32_t)1U << 4U)
 
+#define	SEC_CONF		(0xEE2000B8)
+
+char ek874_board_rev;
+
 static void pfc_reg_write(uint32_t addr, uint32_t data)
 {
 	mmio_write_32(PFC_PMMR, ~data);
@@ -297,6 +302,20 @@ void pfc_init_g2e(void)
 {
 	uint32_t reg;
 
+	/* GP5_19 Pull-up/down function is enabled */
+	mmio_write_32(PFC_PUEN4, mmio_read_32(PFC_PUEN4) | BIT(17));
+	/* GP5_19 Pull-up is enabled */
+	mmio_write_32(PFC_PUD4, mmio_read_32(PFC_PUD4) | BIT(17));
+
+	rzg_micro_delay(10U);
+
+	reg = mmio_read_32(GPIO_INDT5);
+	if (reg & BIT(19))
+		ek874_board_rev = 'C';
+	else
+		ek874_board_rev = 'E';
+
+
 	/* initialize module select */
 	pfc_reg_write(PFC_MOD_SEL0, MOD_SEL0_ADGB_A
 		      | MOD_SEL0_DRIF0_A
@@ -645,4 +664,29 @@ void pfc_init_g2e(void)
 	mmio_write_32(GPIO_INOUTSEL4, 0x00000440U);
 	mmio_write_32(GPIO_INOUTSEL5, 0x00080000U);
 	mmio_write_32(GPIO_INOUTSEL6, 0x00000010U);
+
+	/* initialize setting pfc for using eMMC with SDHI3-IF */
+	if (ek874_board_rev  == 'E') {
+		/*
+		 * Initialize Secure Configuration Register to access QSPI
+		 * when eMMC Boot is performed.
+		 */
+		mmio_write_32(SEC_CONF, 0x0155);
+
+		pfc_reg_write(PFC_GPSR4, GPSR4_SD3_DS
+			      | GPSR4_SD3_DAT7
+			      | GPSR4_SD3_DAT6
+			      | GPSR4_SD3_DAT5
+			      | GPSR4_SD3_DAT4
+			      | GPSR4_SD3_DAT3
+			      | GPSR4_SD3_DAT2
+			      | GPSR4_SD3_DAT1
+			      | GPSR4_SD3_DAT0
+			      | GPSR4_SD3_CMD
+			      | GPSR4_SD3_CLK);
+
+		pfc_reg_write(PFC_PUD3, 0x001FF79FU);
+		pfc_reg_write(PFC_PUEN3, 0x001FF800U);
+		pfc_reg_write(PFC_PUEN4, 0x07800000U);
+	}
 }
diff --git a/plat/renesas/rzg/bl2_plat_setup.c b/plat/renesas/rzg/bl2_plat_setup.c
index 9e4efb5..9866429 100644
--- a/plat/renesas/rzg/bl2_plat_setup.c
+++ b/plat/renesas/rzg/bl2_plat_setup.c
@@ -736,9 +736,15 @@ void bl2_el3_early_platform_setup(u_register_t arg1, u_register_t arg2,
 		break;
 	}
 
-	if (type == BOARD_UNKNOWN || rev == BOARD_REV_UNKNOWN)
+	if (type == BOARD_UNKNOWN || rev == BOARD_REV_UNKNOWN) {
 		NOTICE("BL2: Board is %s Rev.---\n", GET_BOARD_NAME(type));
-	else {
+	} else if (type == BOARD_EK874) {
+		char board_rev;
+
+		board_rev = (char) rev;
+		NOTICE("BL2: Board is %s Rev.%s\n",
+			GET_BOARD_NAME(type), &board_rev);
+	} else {
 		NOTICE("BL2: Board is %s Rev.%d.%d\n",
 		       GET_BOARD_NAME(type),
 		       GET_BOARD_MAJOR(rev), GET_BOARD_MINOR(rev));
-- 
2.7.4

